### ISSUE TITLE
`Qwen3_5MoeTextRotaryEmbedding.inv_freq` reads uninitialized memory after `meta → to_empty(cuda)` materialization

### ISSUE THREAD

**Issue by jamesbraza** [huggingface/transformers#45902]
### System Info

- `transformers` version: 5.6.2
- Platform: Linux-6.8.0-1043-nvidia-x86_64-with-glibc2.35
- Python version: 3.12.13
- Huggingface_hub version: 1.13.0
- Safetensors version: 0.7.0
- Accelerate version: 1.13.0
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.10.0+cu129 (CUDA)
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA H100 80GB HBM3

### Who can help?

@ArthurZucker @Cyrilvallez @zucchini-nlp 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

`Qwen3_5MoeTextRotaryEmbedding.__init__` registers two `persistent=False` buffers, [`inv_freq`](https://github.com/huggingface/transformers/blob/v5.6.2/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L102) and [`original_inv_freq`](https://github.com/huggingface/transformers/blob/v5.6.2/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L103), whose values come from [`rope_init_fn(self.config, device)`](https://github.com/huggingface/transformers/blob/v5.6.2/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L100). When the rotary is constructed on the `meta` device (as SkyRL does [here](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/backends/skyrl_train/workers/model_wrapper.py#L137-L139)) and then materialized to CUDA via `Module.to_empty(device="cuda")`, neither buffer's storage is re-initialized: `to_empty` allocates uninitialized GPU memory, and nothing re-runs `rope_init_fn` against the new device. [`forward`](https://github.com/huggingface/transformers/blob/v5.6.2/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L140) then reads `self.inv_freq` and produces cos/sin from whatever garbage is sitting in that GPU storage.

The first read of `self.inv_freq` therefore returns whatever bytes the caching allocator's free-list happens to hold for that allocation, not the canonical `1.0 / (base ** (arange(0, dim, 2) / dim))` value.

Through an extensive debug script I saw all 32 bins of `inv_freq` returning a consistent `-1.468446e+34` across ref-worker processes, which then overflowed in `inv_freq * position_ids` at `position_id >= 23173`, producing NaN cos/sin and a `policy_kl: nan` downstream in the training loop.

```python
import contextlib
import sys

import torch
from transformers import AutoConfig
from transformers.models.qwen3_5_moe.modeling_qwen3_5_moe import (
    Qwen3_5MoeTextRotaryEmbedding,
)

MODEL = "Qwen/Qwen3.6-35B-A3B"
BIN = 6  # Arbitrary inv_freq index for spot-check
CANONICAL_ABS_TOL = 1e-6
CORRUPTION = -1.468446e34  # Exact fp32 value ref-worker processes saw in inv_freq[*]
# Position ids near the fp32 overflow boundary for inv_freq[6] = -1.468e+34.
# ``inv_freq[6] * position_id`` exceeds fp32 max at position_id >= 23173.
OVERFLOW_BOUNDARY_POSITION = 23173


def summarize(label: str, tensor: torch.Tensor) -> str:
    """Format a one-line summary of a tensor: its device, the value at ``tensor[BIN]``,
    and how many entries hold the production corruption fingerprint.
    """
    val = float(tensor[BIN].item())
    n_corrupt = sum(
        1 for v in tensor.tolist() if abs(v - CORRUPTION) < 1e30 and abs(v) > 1.0
    )
    return (
        f"{label}.device={tensor.device}  {label}[{BIN}]={val:.6e}  "
        f"bins_holding_corruption={n_corrupt}/{tensor.numel()}"
    )


def fwd_overflow(rotary: Qwen3_5MoeTextRotaryEmbedding) -> str:
    """Invoke the rotary's forward on a window of ``position_ids`` straddling the
    fp32 overflow boundary, then reports whether cos/sin went NaN.
    The rotary class itself is upstream at
    https://github.com/huggingface/transformers/blob/v5.8.0/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L86

    Catches ``RuntimeError`` so variant D (where ``inv_freq`` stays on cpu
    while ``x`` is on cuda) reports its mixed-device failure mode
    instead of crashing the whole script.
    """
    x = torch.zeros(1, 1, device="cuda:0", dtype=torch.bfloat16)
    win = 32
    position_ids = torch.arange(
        OVERFLOW_BOUNDARY_POSITION - win,
        OVERFLOW_BOUNDARY_POSITION + win,
        device="cuda:0",
        dtype=torch.long,
    ).unsqueeze(0)
    try:
        with torch.no_grad():
            cos, sin = rotary(x, position_ids)
    except RuntimeError as exc:
        return f"forward: RuntimeError {str(exc).splitlines()[0][:100]}"
    cos_nan = bool(torch.isnan(cos).any().item())
    sin_nan = bool(torch.isnan(sin).any().item())
    cos_at_bin = cos[..., win, BIN].float().mean().item()
    return (
        f"forward: cos.has_nan={cos_nan} sin.has_nan={sin_nan} "
        f"cos[boundary_pos, {BIN}]={cos_at_bin:.4e}"
    )


cfg = AutoConfig.from_pretrained(MODEL, trust_remote_code=False)
text_cfg = cfg.text_config if hasattr(cfg, "text_config") else cfg

# Canonical: what rope_init_fn would compute fresh on cuda
fresh_inv_freq, _ = Qwen3_5MoeTextRotaryEmbedding.compute_default_rope_parameters(
    text_cfg, device=torch.device("cuda:0")
)
canonical = float(fresh_inv_freq[BIN].item())
print(f"=== canonical inv_freq[{BIN}] = {canonical:.10e}\n")

# --- Variant A: direct cuda init --------------------------------------------
rotary_a = Qwen3_5MoeTextRotaryEmbedding(text_cfg, device=torch.device("cuda:0"))
print("--- variant A (direct cuda init):")
print(f"   {summarize('inv_freq', rotary_a.inv_freq)}")
print(f"   {summarize('original_inv_freq', rotary_a.original_inv_freq)}")
print(f"   {fwd_overflow(rotary_a)}\n")

# --- Variant B: meta -> to_empty(cuda) (issue body baseline) ----------------
rotary_b = Qwen3_5MoeTextRotaryEmbedding(text_cfg, device="meta")
rotary_b.to_empty(device="cuda")
print("--- variant B (meta -> to_empty(cuda)):")
print(f"   {summarize('inv_freq', rotary_b.inv_freq)}")
print(f"   {summarize('original_inv_freq', rotary_b.original_inv_freq)}")
print(f"   {fwd_overflow(rotary_b)}\n")

# --- Variant C: cuda alloc pre-fill -> meta -> to_empty(cuda) ---------------
print("--- variant C (cuda alloc pre-fill -> meta -> to_empty(cuda)):")
for n_elems in (32, 64, 128, 256):
    for _ in range(8):
        _j = torch.full((n_elems,), CORRUPTION, device="cuda:0", dtype=torch.float32)
        del _j
    rotary_c = Qwen3_5MoeTextRotaryEmbedding(text_cfg, device="meta")
    rotary_c.to_empty(device="cuda")
    print(f"   junk_size={n_elems:>4}  {summarize('inv_freq', rotary_c.inv_freq)}")
print(f"   {fwd_overflow(rotary_c)}\n")

# --- Variant D: meta + direct _buffers[...] = empty(cpu) (bypasses _apply) --
# https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/backends/skyrl_train/distributed/fsdp_utils.py#L261-L279
rotary_d = Qwen3_5MoeTextRotaryEmbedding(text_cfg, device="meta")
rotary_d._buffers["inv_freq"] = torch.empty(32, dtype=torch.float32, device="cpu")
rotary_d._buffers["original_inv_freq"] = torch.empty(
    32, dtype=torch.float32, device="cpu"
)
print("--- variant D (meta + direct _buffers[...] = empty(cpu), bypasses _apply):")
print(f"   {summarize('inv_freq', rotary_d.inv_freq)}")
print(f"   {summarize('original_inv_freq', rotary_d.original_inv_freq)}")
print(f"   {fwd_overflow(rotary_d)}\n")

print("=== expected outcomes under PR #45903 (already merged in main) ===")
print("   A: canonical, no nan.")
print("   B: canonical via the _apply hook fired by to_empty.")
print("   C: canonical via the _apply hook (cuda pre-fill bytes overwritten).")
print("   D: PR #45903's _apply hook does NOT fire (direct _buffers assignment")
print("      bypasses it). inv_freq holds whatever bytes the host allocator")
print("      returned; the forward consuming the buffer at position_id >= 23173")
print("      produces wrong cos/sin (NaN if the bytes overflow fp32, else")
print("      structurally wrong values like cos=1 from zero-filled pages).")
```

Output:

```none
--- variant A (direct cuda init):
   inv_freq.device=cuda:0  inv_freq[6]=4.869675e-02
   forward: cos.has_nan=False sin.has_nan=False cos[boundary_pos, 6]=-8.1641e-01

--- variant B (meta -> to_empty(cuda)):
   inv_freq.device=cuda:0  inv_freq[6]=4.869675e-02
   forward: cos.has_nan=False sin.has_nan=False cos[boundary_pos, 6]=-8.1641e-01

--- variant C (cuda alloc pre-fill -> meta -> to_empty(cuda)):
   junk_size=  32  inv_freq.device=cuda:0  inv_freq[6]=4.869675e-02
   ...
   forward: cos.has_nan=False sin.has_nan=False cos[boundary_pos, 6]=-8.1641e-01

--- variant D (meta + direct _buffers[...] = empty(cpu), bypasses _apply):
   inv_freq.device=cpu  inv_freq[6]=8.407791e-45
   forward: cos.has_nan=False sin.has_nan=False cos[boundary_pos, 6]=1.0000e+00
```

So we see:
- Variant A's direct-on-cuda init is correct on both buffers.
- Variant B shows the bug clearly without any pre-filling: `original_inv_freq[6] = 0.0`, not the canonical `0.0487`.
    - The `inv_freq[6] = 0.0487` reading is coincidental aliasing: the allocator's free-list happens to hold bytes from the canonical computation we did three lines earlier.
- Variant C confirms the read-uninitialized-memory mechanism by pre-filling the allocator's small-bucket pool with a specific fp32 pattern and showing it survives into the materialized buffer at every bucket size tried.
  - Variant D shows the bypass that _apply does not catch: module._buffers["inv_freq"] = empty(cpu) directly (what SkyRL's _sync_non_persistent_buffers does at FSDP2 init) routes around
   Module._apply / register_buffer / __setattr__, so PR #45903's hook never fires. inv_freq.device = cpu and inv_freq[6] holds uninitialized bytes; the silent-corruption fingerprint
  here is cos[boundary_pos, 6] = 1.0 instead of the canonical -0.8164 — in production we hit inv_freq[*] = -1.468e+34 and fp32 overflow → NaN cos/sin → policy_kl: nan.
- Variant D simulates the FSDP2 helper pattern (e.g. [SkyRL's `_sync_non_persistent_buffers`](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/backends/skyrl_train/distributed/fsdp_utils.py#L261-L279)) that writes directly into `module._buffers["inv_freq"]` after construction; `transformers` does not detect the bypass, `cos[boundary_pos, 6] = 1.0` from zero-filled pages rather than the canonical `-0.8164`, and the `inv_freq` overflows fp32 to NaN cos/sin at `position_id >= 23173`.

### Expected behavior

`Qwen3_5MoeTextRotaryEmbedding(cfg, device="meta").to_empty(device="cuda")` should produce a rotary embedding whose `inv_freq` and `original_inv_freq` match the canonical `rope_init_fn(cfg, device=cuda)` values, and `forward` should produce canonical cos/sin even after FSDP2 helpers reassign `module._buffers[...]` directly (variant D above). As-is, both buffers carry the caching allocator's free-list bytes and `forward` consumes them verbatim. The result is silent corruption when those bytes are zero (`cos = 1.0` from `cos(0)`), or NaN cos/sin when they overflow fp32.

**Comment by jshaofa-ui** [comment]
## 🔧 Solution Proposal: Qwen3_5MoeTextRotaryEmbedding Uninitialized Memory After `to_empty()`

### Root Cause

`Qwen3_5MoeTextRotaryEmbedding.__init__` registers two `persistent=False` buffers (`inv_freq` and `original_inv_freq`) whose values come from `rope_init_fn(self.config, device)`. When the rotary is constructed on the `meta` device and then materialized to CUDA via `Module.to_empty(device="cuda")`, `to_empty` allocates uninitialized GPU memory without re-running `rope_init_fn`. The `forward()` method then reads garbage values from `self.inv_freq`, producing NaN cos/sin tensors.

### Fix

Override `to_empty` in `Qwen3_5MoeTextRotaryEmbedding` to re-initialize the buffers after materialization:

```python
# In src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py

class Qwen3_5MoeTextRotaryEmbedding(nn.Module):
    def __init__(self, config: Qwen3_5MoeTextConfig, device=None):
        super().__init__()
        # ... existing code ...
        self.rope_init_fn = lambda cfg, dev: compute_default_rope_parameters(cfg, device=dev)
        inv_freq, original_inv_freq = self.rope_init_fn(config, device)
        self.register_buffer("inv_freq", inv_freq, persistent=False)
        self.register_buffer("original_inv_freq", original_inv_freq, persistent=False)

    def to_empty(self, device, *, copy: bool = False):
        """Override to re-initialize inv_freq buffers after materialization from meta device."""
        super().to_empty(device, copy=copy)
        # Re-initialize the rotary frequency buffers on the target device
        inv_freq, original_inv_freq = self.rope_init_fn(self.config, device)
        self.inv_freq.copy_(inv_freq)
        self.original_inv_freq.copy_(original_inv_freq)
        return self
```

### Files to Modify
- `src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py` — Add `to_empty` override

### Regression Test
```python
def test_qwen3_5_moe_rotary_meta_to_empty():
    cfg = AutoConfig.from_pretrained("Qwen/Qwen3.6-35B-A3B", trust_remote_code=False)
    text_cfg = cfg.text_config

    # Construct on meta, materialize to CUDA
    rotary = Qwen3_5MoeTextRotaryEmbedding(text_cfg, device="meta")
    rotary.to_empty(device="cuda")

    # Verify inv_freq is properly initialized (not garbage)
    canonical, _ = Qwen3_5MoeTextRotaryEmbedding.compute_default_rope_parameters(
        text_cfg, device=torch.device("cuda")
    )
    assert torch.allclose(rotary.inv_freq, canonical), "inv_freq not re-initialized after to_empty"
    assert not torch.isnan(rotary.inv_freq).any(), "inv_freq contains NaN"

    # Verify forward produces valid cos/sin
    position_ids = torch.arange(4096, device="cuda")
    cos, sin = rotary(position_ids)
    assert not torch.isnan(cos).any(), "cos contains NaN"
    assert not torch.isnan(sin).any(), "sin contains NaN"
```

### Impact
- Fixes NaN loss in training pipelines using SkyRL or similar frameworks that construct models on meta device
- No breaking changes — `to_empty` override is additive
- Minimal code change (single method override)

**Comment by Rocketknight1** [comment]
@jshaofa-ui stop spamming code agent output into random issues. It's not helpful and it just creates noise

**Comment by Cyrilvallez** [comment]
Looks like your issue comes from the external library you mentionned (SkyRL)! It's expected that the module will not hold any real correct values after meta -> device change. In transformers, we call `init_weights` to avoid this issue.
Feel free to mreopen if you think this is not solved! 🤗

**Comment by sorayamoreau48-a11y** [comment]
Great attention to detail in this PR
